CONTRIBUTING.md

```diff
@@ -14,3 +14,223 @@ If you are sending a pull request and it's a non-trivial change beyond fixing ty
+
+- Avoid using redundant keywords. To list a few:
+  - `final` method modifier in a `final` class
+  - `static` or `public` modifier in an `interface`
+  - `public` method modifier in a package-local or private class
+  - `private` constructor modifier in an `enum`
+  - field access prefixed with `this.` where unnecessary
+
+- Organize class members carefully for readability, using **top-down** approach.
+  Although there's no absolute rule of thumb, it's usually like:
+  - `static` fields
+  - `static` methods
+  - member fields
+  - constructors
+  - member methods
+  - utility methods (both `static` and member)
+  - inner classes
+
+- Do explicit `null`-check on the parameters of user-facing public methods.
+  Always use `Objects.requireNonNull(Object, String)` to do a `null`-check.
+
+  ```java
+  import static java.util.Objects.requireNonNull;
+
+  public void setName(String name) {
+    this.name = requireNonNull(name, "name");
+  }
+  ```
+
+  If you are using IntelliJ IDEA and you imported the `settings.jar` as explained
+  above, try the live template `rnn` and `rnna` which will save a lot of time.
+
+- Avoid unnecessary `null`-checks, including the hidden checks in `Objects.hashCode()` and `Objects.equals()`.
+
+  ```java
+  public final class MyClass {
+    private final String name;
+
+    public MyClass(String name) {
+        // We are sure 'name' is always non-null.
+        this.name = requireNonNull(name, "name");
+    }
+
+    @Override
+    public int hashCode() {
+        // OK
+        return name.hashCode();
+        // Not OK
+        return Objects.hash(name);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
+        if (getClass() != obj.getClass()) {
+            return false;
+        }
+
+        // OK
+        return name.equals(((MyClass) obj).name);
+        // Not OK
+        return Objects.equals(name, ((MyClass) obj).name);
+    }
+  }
+  ```
+
+- Do explicit validation on the parameters of user-facing public methods.
+  When raising an exception, always specify the detailed message in the following
+  format:
+
+  ```java
+  public void setValue(int value) {
+      if (value < 0) {
+          // Note that the exception message contains the value of the offending
+          // parameter as well as the expected value.
+          throw new IllegalArgumentException("value: " + value + " (expected: >= 0)");
+      }
+  }
+  ```
+
+- Prefer using plain JDK API when the same behavior can be achieved with the same
+  amount of code.
+
+  ```java
+  // Prefer A
+  Map<String, String> map = new HashMap<>();   // A (JDK)
+  Map<String, String> map = Maps.newHashMap(); // B (Guava)
+
+  // Prefer B
+  List<String> list = Collections.unmodifiableList(  // A (JDK)
+          otherList.stream().filter(...).collect(Collectors.toList()));
+  List<String> list = otherList.stream().filter(...) // B (Guava)
```

Comment:
ogh, I also missed "// Prefer B" and took in opposite meaning.
Would better write something like "// B (Guava), better as it's much simpler".